plugins/header_rewrite/operators.cc

```diff
@@ -1219,3 +1304,60 @@ OperatorRunPlugin::exec(const Resources &res) const
     _plugin->doRemap(res.txnp, res._rri);
   }
 }
+
+// OperatorSetBody
+void
+OperatorSetBodyFrom::initialize(Parser &p)
+{
+  Operator::initialize(p);
+  // we want the arg since body only takes one value
+  _value.set_value(p.get_arg());
+  require_resources(RSRC_SERVER_RESPONSE_HEADERS);
+  require_resources(RSRC_RESPONSE_STATUS);
+}
+
+void
+OperatorSetBodyFrom::initialize_hooks()
+{
+  add_allowed_hook(TS_HTTP_READ_RESPONSE_HDR_HOOK);
+}
+
+void
+OperatorSetBodyFrom::exec(const Resources &res) const
+{
+  if (TSHttpTxnIsInternal(res.txnp)) {
+    // If this is triggered by an internal transaction, a infinte loop may occur
+    // It should only be triggered by the original transaction sent by the client
+    Dbg(pi_dbg_ctl, "OperatorSetBodyFrom triggered by an internal transaction");
+    return;
+  }
+
+  char req_buf[MAX_SIZE];
+  int  req_buf_size = 0;
+  if (createRequestString(_value.get_value(), req_buf, &req_buf_size) == TS_SUCCESS) {
+    TSCont fetchCont = TSContCreate(handleFetchEvents, TSMutexCreate());
+    TSContDataSet(fetchCont, static_cast<void *>(res.txnp));
+
+    TSHttpTxnHookAdd(res.txnp, TS_HTTP_SEND_RESPONSE_HDR_HOOK, fetchCont);
```

Comment:
Suppose there is another continuation for the transaction, that is either after the continuation calling this exec func on the READ_RESPONSE_HDR_HOOK, or before fetchCont on the SEND_RESPONSE_HDR_HOOK  And, suppose this continuation, like fetchEvent for the SEND_RESPONSE_HDR evert, does not call TxnReenable().  Seems like there could be a race condition, where the URL fetch body done event reenables for that other continuation by mistake.

Could you add a return value to the operator exec funtions, which would control whether the continuation running the exec functions called TxnReenable()?  That way, no other continuations of transaction hooks would be run until the fetch of the body URL finished.

Reply:
I see what you are saying but I dont see how the exec func returning something would help. I think a return value would only have impact on header_rewrite rather than all plugins.

Reply:
This is what I'm proposing, for the exec() function to return a flag that controls whether reenable is called:  https://github.com/ywkaras/trafficserver/commit/0ef4d24b905953f88d866b3c11cd5cfb88e55f89

Unfortunately, for this approach to work when the operator is run in the remap pseudo hook, the call to TSFetchUrl() would have to be deferred to the POST_REMAP hook.

Reply:
I see. I think it would be better to seperate these changes into 2 PRs because the change you are proposing impacts all of header rewrite.

Reply:
OK, I can put up a follow-on PR once this is merged.